runtime/src/main/cpp/Atomic.cpp

```diff
@@ -62,9 +82,38 @@ void Kotlin_AtomicReference_checkIfFrozen(KRef value) {
     }
 }
 
-KRef Kotlin_AtomicReference_compareAndSwap(KRef thiz, KRef expectedValue, KRef newValue) {
+OBJ_GETTER(Kotlin_AtomicReference_compareAndSwap, KRef thiz, KRef expectedValue, KRef newValue) {
     Kotlin_AtomicReference_checkIfFrozen(newValue);
-    return compareAndSwapImpl(thiz, expectedValue, newValue);
+    if (expectedValue == newValue) {
+        RETURN_OBJ(expectedValue);
+    }
+    // See Kotlin_AtomicReference_get() for explanations, why locking is needed.
+    lock(thiz);
+    // As we lock [thiz], atomicity here is not really important.
+    KRef old = compareAndSwapImpl(thiz, expectedValue, newValue);
```

Comment:
Since we use lock, it is not necessary to perform atomic CAS here.

Reply:
Yes, as said in comment, it is just shorter.